GH actions for rendering of README.Rmd

jibarozzo opened the pull request:
Included `render-markdown.yaml` through `usethis::use_github_action("render-rmarkdown")` default settings. I added dummy test in .Rmd

jibarozzo commented:
This PR addresses #16. It triggers on every PR and on `main`. All checks pass.

jdhoffa commented on .github/workflows/render-rmarkdown.yaml:
```diff
@@ -0,0 +1,68 @@
+# Need help debugging build failures? Start at https://github.com/r-lib/actions#where-to-find-help
+on:
+  push:
+    branches: [main, master]
```
NB: I don't imagine we will start having a `master` branch in the future
```suggestion
    branches: [main]
```

jdhoffa commented on .github/workflows/render-rmarkdown.yaml:
```diff
@@ -0,0 +1,68 @@
+      - name: Render RMarkdown files
+        shell: bash
+        run: |
+          RMD_PATH=$(git diff --name-only ${{ github.event.before }} ${{ github.sha }} | grep '\.Rmd$' || true)
```
This is awesome, but just a note here:
``` bash
RMD_PATH=$(git diff --name-only ${{ github.event.before }} ${{ github.sha }} | grep '\.Rmd$' || true)
```

This bash command will run a `git diff` and `grep` any `.Rmd` files that have been changed before/ after the PR. While that will catch any direct changes to the `.Rmd` itself, it _won't_ catch underlying changes in the `blazr` package. 

So for example, if we were to update the functionality of a `blazr` function but didn't make any changes to the `README`, this action wouldn't catch those changes, and wouldn't try to trigger a re-knit of the `README.Rmd`. 

Certainly not a blocking comment to this PR, but maybe something worth expanding on at a later date to improve the action

jdhoffa replied:
Suggest to extract this comment into a new tracking issue

jibarozzo replied:
Created new issue #18

jdhoffa commented on README.Rmd:
````diff
@@ -43,3 +43,11 @@ library(blazr)
 
 blazr::to_upper("hello, world")
 ```
+
+Testing dummy example with GH actions Rmd render
+
+```{r example2}
+library(blazr)
+
+blazr::to_upper("This is a test ... again")
+```
````
Please revert this testing commit 😊 
```suggestion
```

jdhoffa reviewed (changes requested):
Some knit comments, one suggestion.

jdhoffa commented on .github/workflows/render-rmarkdown.yaml:
```diff
@@ -0,0 +1,68 @@
+            echo "No RMarkdown files changed."
+          fi
+
+      - name: Commit Results
+        shell: bash
+        run: |
+          if [ -n "$RMD_PATH" ]; then
+            MD_FILES=$(echo "$RMD_PATH" | sed 's/\.Rmd$/\.md/')
+            git config --local user.name "$GITHUB_ACTOR"
+            git config --local user.email "$GITHUB_ACTOR@users.noreply.github.com"
+            git add $MD_FILES
+            git commit -m 'Re-build RMarkdown files' || echo "No changes to commit"
+            git push origin || echo "No changes to commit"
+          else
+            echo "No Markdown files to commit."
+          fi
```
Is this part of the action actually ever triggering? 

I would expect a commit with the message `Re-build RMarkdown files`, but I don't actually see one in the commit log for this PR.

jibarozzo replied:
This, I didn't change that much from the [initial template](https://github.com/r-lib/actions/tree/v2/examples#render-rmarkdown). I'll look into this more.

jibarozzo replied:
As I understand, we can make the commit in the current branch then merge or in `main` when we merge it. How it is set up now it would knit when merged with `main`. It is preferable to commit to current branch then merge into `main`. I'd thought about it but didn't make it happen.  I'll merge once I resolve this.

jibarozzo replied:
See discussion below.

jibarozzo commented:
This PR's last two commits succesfully detect changes in `.Rmd` files and trigger the re-build of README.md. 

### Changes from last review.  

- Includes new job **Showing GH context refs**: This allows for debugging and tracing knowing where things are being committed. 
- Added `env` variable `BRANCH_NAME`
-  Checkout PR branch job targets ${{ github.ref  }}: Makes sure we checkout to current PR branch, no a detached head which is the default.
- Fixed typo in `README.Rmd` that hindered installation of `blazr` for example
-  Added `git push origin HEAD:${{ github.head_ref }}`: This is commits to the current PR branch, because even after checkout it is still a 'headless' environment that it is working on. Based on error results from commit [845f0f5](https://github.com/r-staceans/blazr/pull/17/commits/845f0f56abe92aa37c50c887265c2b3e767f57e3) (see **Commit Results** section in [test](https://github.com/r-staceans/blazr/actions/runs/12305143063/job/34344104307)). Commit [09dac1f](https://github.com/r-staceans/blazr/pull/17/commits/09dac1fffe5c8272b7aa530726d38d89af0b0df3) correctly addresses this issue. 

### Todos

- [ ] Trigger of README after changes to `blazr` features and functionalities. Logic for this will be handled in another PR See issue #18

jdhoffa commented on README.Rmd:
````diff
@@ -29,7 +29,7 @@ The goal of blazr is to provide R with a blazingly fast interface to multi-threa
 Install the development version of `blazr` from GitHub with:
 
-``` r
+```{r}
````
I think this is not actually desirable. That code chunk was only meant to be a demo to users on how they _might_ want to install the package, not necessarily meant to run every time we knit the README.


I don't think we want the package to re-install from GitHub every-time we re-knit the README locally. (Consider if we are knitting the `README.Rmd` on a feature branch, it would be impossible/ very difficult to see those changes, as `pak` would always re-install from `main` on GitHub)

jibarozzo replied:
I see where you're coming from. I'll revert this and add the package through the dependencies job.

jdhoffa commented on .github/workflows/render-rmarkdown.yaml:
```diff
@@ -0,0 +1,78 @@
+      - name: Setup Pandoc
+        uses: r-lib/actions/setup-pandoc@v2
+
+      - name: Setup R Dependencies
+        uses: r-lib/actions/setup-r-dependencies@v2
+        with:
+          extra-packages: any::rmarkdown
```
As I understand it, this step takes care of installing the local `blazr` package.
See: https://github.com/r-lib/actions/tree/v2-branch/setup-r-dependencies

jibarozzo replied:
This is probably handled 
- `uses: r-lib/actions/setup-r-dependencies@v2
  with:
    extra-packages: any::rmarkdown, local::.`
    

Note to self: See [here](https://github.com/r-lib/actions/blob/v2-branch/examples%2FREADME.md)

jibarozzo replied:
Resolved. See commit message below.

jdhoffa commented on .github/workflows/render-rmarkdown.yaml:
```diff
@@ -0,0 +1,78 @@
+            git config --local user.email "$GITHUB_ACTOR@users.noreply.github.com"
+            git add $MD_FILES
+            git commit -m 'Re-build RMarkdown files' || echo "Re-build changes to commit"
+            git push origin HEAD:${{ github.head_ref }} || echo "Changes pushed to ${{ github.head_ref }} to commit"
```
This is awesome!!

jdhoffa reviewed (changes requested):
A note about the changes to `README`, but the action itself looks awesome!!
Hell yeah, thank you 😊

jdhoffa commented on README.Rmd:
````diff
@@ -29,7 +29,7 @@ The goal of blazr is to provide R with a blazingly fast interface to multi-threa
 Install the development version of `blazr` from GitHub with:
 
-``` r
+```r
````
```suggestion
``` r
```

jdhoffa commented on README.md:
````diff
@@ -24,6 +24,37 @@ Install the development version of `blazr` from GitHub with:
 ``` r
 # install.packages("pak")
 pak::pak("r-staceans/blazr")
+#> 
+#> → Will install 1 package.
+#> → Will download 1 package with unknown size.
+#> + blazr   0.0.0.9000 [bld][cmp][dl] (GitHub: ac3d506) + ✔ rustc, ✔ cargo
+#> ✔ All system requirements are already installed.
+#> 
+#> ℹ Getting 1 pkg with unknown size
+#> ✔ Got blazr 0.0.0.9000 (source) (1.73 MB)
+#> ℹ Installing system requirements
+#> ℹ Executing `sudo sh -c apt-get -y update`
+#> Get:1 file:/etc/apt/apt-mirrors.txt Mirrorlist [142 B]
+#> Hit:2 http://azure.archive.ubuntu.com/ubuntu noble InRelease
+#> Hit:3 http://azure.archive.ubuntu.com/ubuntu noble-updates InRelease
+#> Hit:4 http://azure.archive.ubuntu.com/ubuntu noble-backports InRelease
+#> Hit:5 http://azure.archive.ubuntu.com/ubuntu noble-security InRelease
+#> Hit:6 https://packages.microsoft.com/repos/azure-cli noble InRelease
+#> Hit:7 https://packages.microsoft.com/ubuntu/24.04/prod noble InRelease
+#> Reading package lists...
+#> ℹ Executing `sudo sh -c apt-get -y install rustc cargo`
+#> Reading package lists...
+#> Building dependency tree...
+#> Reading state information...
+#> rustc is already the newest version (1.75.0+dfsg0ubuntu1-0ubuntu7.1).
+#> cargo is already the newest version (1.75.0+dfsg0ubuntu1-0ubuntu7.1).
+#> 0 upgraded, 0 newly installed, 0 to remove and 79 not upgraded.
+#> ℹ Packaging blazr 0.0.0.9000
+#> ✔ Packaged blazr 0.0.0.9000 (590ms)
+#> ℹ Building blazr 0.0.0.9000
+#> ✔ Built blazr 0.0.0.9000 (7.8s)
+#> ✔ Installed blazr 0.0.0.9000 (github::r-staceans/blazr@ac3d506) (65ms)
+#> ✔ 1 pkg: added 1, dld 1 (NA B) [13.5s]
````
```suggestion
```

jibarozzo replied:
Solved with [070061f](https://github.com/r-staceans/blazr/pull/17/commits/070061fd258552852e577385a8fedfff45b1dd4a)

jdhoffa reviewed (changes requested):
Two suggestions to get the `README.*` files back to 0 diff 😊

jdhoffa reviewed (approved):
LGTM 🎉
